File: manual/src/test/java/examples/java/FactDocTest.java
package examples.java;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.YearMonth;
import java.util.UUID;
import org.junit.jupiter.api.Test;
import org.shakespeareframework.Actor;
import org.shakespeareframework.Fact;

class FactDocTest {

  @Test
  void act1() {
    // tag::create-actor[]
    var kate = new Actor("Kate");
    // end::create-actor[]
    // tag::learn-fact[]
    kate.learns(new PhoneNumber("+49 0180 4 100 100"));
    // end::learn-fact[]
    // tag::remember-fact[]
    PhoneNumber katesPhoneNumber = kate.remembers(PhoneNumber.class);
    // end::remember-fact[]
    assertThat(katesPhoneNumber).isInstanceOf(PhoneNumber.class);
    // tag::relearn-fact[]
    kate.learns(new PhoneNumber("+1 (303) 499-7111"));
    // end::relearn-fact[]
    // tag::remember-changed-fact[]
    PhoneNumber katesNewPhoneNumber = kate.remembers(PhoneNumber.class);
    assertThat(katesNewPhoneNumber).isEqualTo(new PhoneNumber("+1 (303) 499-7111"));
    // end::remember-changed-fact[]
  }

  // tag::fact[]
  record PhoneNumber(String number) implements Fact {}

  // end::fact[]

  @Test
  void act2() {
    var robin = new Actor("Robin");
    // tag::learn-poly-fact[]
    robin.learns(new PhoneNumbers("+49 0180 4 100 100", "+1 (303) 499-7111"));
    // end::learn-poly-fact[]
    // tag::remember-poly-fact[]
    var robinsPhoneNumbers = robin.remembers(PhoneNumbers.class);
    var robinsHomePhoneNumber = robinsPhoneNumbers.home();
    var robinsWorkPhoneNumber = robinsPhoneNumbers.work();
    // end::remember-poly-fact[]
    assertThat(robinsPhoneNumbers).isInstanceOf(PhoneNumbers.class);
    assertThat(robinsHomePhoneNumber).isNotBlank();
    assertThat(robinsWorkPhoneNumber).isNotBlank();
  }

  // tag::poly-fact[]
  record PhoneNumbers(String home, String work) implements Fact {}

  // end::poly-fact[]

  @Test
  void act3() {
    var bill = new Actor("Bill");
    // tag::learn-default-fact
    bill.learns(CreditCard.DEFAULT_VISA);
    // end::learn-default-fact
    assertThat(bill.remembers(CreditCard.class)).isEqualTo(CreditCard.DEFAULT_VISA);
    // tag::learn-generated-fact
    bill.learns(EmailAddress.random());
    // end::learn-generated-fact
    assertThat(bill.remembers(EmailAddress.class)).isNotEqualTo(EmailAddress.random());
  }

  // tag::fact-with-default[]
  record CreditCard(String type, String number, YearMonth expiration) implements Fact {
    public static CreditCard DEFAULT_VISA =
        new CreditCard("visa", "4111111111111111", YearMonth.of(2026, 10));
  }

  // end::fact-with-default[]

  // tag::fact-with-generator[]
  record EmailAddress(String address) implements Fact {
    public static EmailAddress random() {
      return new EmailAddress("%s@shakespeareframework.org".formatted(UUID.randomUUID()));
    }
  }
  // end::fact-with-generator[]

}

File: manual/src/test/kotlin/examples/kotlin/FactDocTest.kt
package examples.kotlin

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.shakespeareframework.Actor
import org.shakespeareframework.Fact

import java.time.YearMonth
import java.util.UUID

class FactDocTest {

  @Test
  fun act1() {
    // tag::create-actor[]
    val kate = Actor("Kate")
    // end::create-actor[]
    // tag::learn-fact[]
    kate.learns(PhoneNumber("+49 0180 4 100 100"))
    // end::learn-fact[]
    // tag::remember-fact[]
    val katesPhoneNumber = kate.remembers(PhoneNumber::class.java)
    // end::remember-fact[]
    assertThat(katesPhoneNumber).isInstanceOf(PhoneNumber::class.java)
    // tag::relearn-fact[]
    kate.learns(PhoneNumber("+1 (303) 499-7111"))
    // end::relearn-fact[]
    // tag::remember-changed-fact[]
    val katesNewPhoneNumber = kate.remembers(PhoneNumber::class.java)
    assertThat(katesNewPhoneNumber).isEqualTo(PhoneNumber("+1 (303) 499-7111"))
    // end::remember-changed-fact[]
  }

  // tag::fact[]
  data class PhoneNumber(val number: String) : Fact

  // end::fact[]

  @Test
  fun act2() {
    val robin = Actor("Robin")
    // tag::learn-poly-fact[]
    robin.learns(PhoneNumbers(home = "+49 0180 4 100 100", work = "+1 (303) 499-7111"))
    // end::learn-poly-fact[]
    // tag::remember-poly-fact[]
    val robinsPhoneNumbers = robin.remembers(PhoneNumbers::class.java)
    val robinsHomePhoneNumber = robinsPhoneNumbers.home
    val robinsWorkPhoneNumber = robinsPhoneNumbers.work
    // end::remember-poly-fact[]
    assertThat(robinsPhoneNumbers).isInstanceOf(PhoneNumbers::class.java)
    assertThat(robinsHomePhoneNumber).isNotBlank
    assertThat(robinsWorkPhoneNumber).isNotBlank
  }

  // tag::poly-fact[]
  data class PhoneNumbers(val home: String, val work: String) : Fact

  // end::poly-fact[]

  @Test
  fun act3() {
    val bill = Actor("Bill")
    // tag::learn-default-fact
    bill.learns(CreditCard.DEFAULT_VISA)
    // end::learn-default-fact
    assertThat(bill.remembers(CreditCard::class.java)).isEqualTo(CreditCard.DEFAULT_VISA)
    // tag::learn-generated-fact
    bill.learns(EmailAddress.random())
    // end::learn-generated-fact
    assertThat(bill.remembers(EmailAddress::class.java)).isNotEqualTo(EmailAddress.random())
  }

  // tag::fact-with-default[]
  data class CreditCard(val type: String, val number: String, val expiration: YearMonth) : Fact {
    companion object {
      val DEFAULT_VISA =
        CreditCard("visa", "4111111111111111", YearMonth.of(2026, 10))
    }
  }

  // end::fact-with-default[]

  // tag::fact-with-generator[]
  data class EmailAddress(val address: String) : Fact {
    companion object {
      fun random() = EmailAddress("${UUID.randomUUID()}@shakespeareframework.org")
    }
  }
  // end::fact-with-generator[]

}
